docs/design/ConstructiveSolidGeometry.MD

```diff
@@ -0,0 +1,136 @@
+ - Enter the right (union) node.
+ - Enter the left green cylinder. It's a leaf node so we leave it immediately and add it to the list (4. Green cylinder)
+ - Enter the second right (union) node.
+ - Enter the left green cylinder. It's a leaf node so we leave it immediately and add it to the list (5. Green cylinder)
+ - Enter the right green cylinder. It's a leaf node so we leave it immediately and add it to the list (6. Green cylinder)
+ - Leave the second right (union) node and add it to the list (7. Union)
+ - Leave the first right (union) node and add it to the list (8. Union)
+ - Leave the root (subtraction) node and add it to the list (9. Subtraction)
+
+The list now looks like this:
+```
+1. Red cube
+2. Blue sphere
+3. Intersect
+4. Green cylinder
+5. Green cylinder
+6. Green cylinder
+7. Union
+8. Union
+9. Subtraction
+```
+If we treat it as a Reverse Polish Notation we can iterate it using a simple stack.
+
+```
+1. -> [1. RC]
+2. -> [1. RC, 2. BS]
+3. -> [1. RC * BS]
+4. -> [1. RC * BS, 2. GC]
+5. -> [1. RC * BS, 2. GC, 3. GC]
+6. -> [1. RC * BS, 2. GC, 3. GC, 4. GC]
+7. -> [1. RC * BS, 2. GC, 3. GC + GC]
+8. -> [1. RC * BS, 2. GC + GC + GC]
+9. -> [1. RC * BS - (GC + GC + GC)]
+```
+
+## The Shader
+
+We will use the following structure to represent a flattened Node in a list:
+```
+struct NodeCommand
+{
+    float4  m_PosSize;    // Position in 3D (xyz) and a uniform scale
+    float4  m_AuxData1;   // Auxiliary data
+    float4  m_AuxData2;   // Auxiliary data
+    uint4   m_Control;    // OpCode, AuxCode1, AuxCode2, ControlCode
+}
+```
+
+To keep nodes uniform, they are all the same size (64 bytes = float4x4) where the first entry is always the node's position and size.
+
+`m_PosSize.xyz` - position in 3D.
+
+`m_PosSize.w` - uniform scale. Non-uniform scaling can be added either
+
+`m_AuxData1` - auxiliary data for the command. What data is encoded here is specified by the `AuxCode1` code (`m_Control.y`)
+
+`m_AuxData2` - auxiliary data for the command. What data is encoded here is specified by the `AuxCode2` code (`m_Control.z`). Both auxiliary codes use the same list of codes and it's up to the user to decide which slot to use, but if both specify the same data it's undefined which one should be used (it's up to the implementation).
+
+`OpCode` (`m_Control.x`) - defines what the node does.
+
+`ControlCode` (`m_Control.w`) - control flags used by the shader.
+
+### Examples of some of the codes
+
+`OpCodeSphere` - the current node should trace a sphere using `sdSphere` and put the distance on the stack.
+
+`AuxCodeOpData` - the auxiliary data is operator data. Operator data is different per OpCode. For example, the `OpCodeSphere` requires a radius for the sphere it traces. If OpData is available, it gets it from GetOpData().x;
+
+`AuxCodeRotation` - the auxiliary data is rotation data for the node, given as a quaternion. See the Math section for details.
+
+## The Math
+
+Location: `//.../Shaders/Math.h`
```

Comment:
Not sure if this `Math.h` is added to the branch repo already, but can't find it.

Reply:
Ok, sure.